spec/std/openssl/ssl/hostname_validation_spec.cr

```diff
@@ -0,0 +1,54 @@
+require "spec"
+require "openssl/ssl/hostname_validation"
+
+describe OpenSSL::SSL::HostnameValidation do
+  describe "matches_hostname?" do
+    it "matches ipv4, ipv6" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.1", "192.168.1.1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.2", "192.168.1.1").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::2").should be_false
+    end
+
+    it "won't wildcard match against ip" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("*.168.1.1", "192.168.1.1").should be_false
+    end
+
+    it "skips trailing dot" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com.", "example.com").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", "example.com.").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?(".example.com", "example.com").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", ".example.com").should be_false
```

Comment:
SIP also demands to prefer a `URI` SAN with the `sip` scheme over a `DNS` SAN. They also forbid to match IDNs. Tbh RFC5922 reads like a mixture of paranoia and implementation laziness. RFC6125 is more recent and IMO has done a better job of actually trying to define strict rules for common implementation behavior, instead of just dodging the issue altogether.

However of course we should add API to allow the user to disable wildcard matching entirely, but I'd do so in a separate PR as it affects both code paths, the ones for OpenSSL 1.0.2+ and below.